src/Algebra/Module.agda

```diff
@@ -0,0 +1,283 @@
+
+  +ᴹ-commutativeMonoid : CommutativeMonoid m ℓm
+  +ᴹ-commutativeMonoid =
+    record { isCommutativeMonoid = +ᴹ-isCommutativeMonoid }
+
+  open CommutativeMonoid +ᴹ-commutativeMonoid public
+    using ()
+    renaming ( monoid to +ᴹ-monoid; semigroup to +ᴹ-semigroup
+             ; magma to +ᴹ-magma; rawMagma to +ᴹ-rawMagma
+             ; rawMonoid to +ᴹ-rawMonoid)
+
+record RightModule {r ℓr} (ring : Ring r ℓr) m ℓm
+                   : Set (r ⊔ ℓr ⊔ suc (m ⊔ ℓm)) where
+  open Ring ring
+
+  infixr 8 -ᴹ_
+  infixl 7 _*ᵣ_
+  infixl 6 _+ᴹ_
+  infix 4 _≈ᴹ_
+
+  field
+    Carrierᴹ : Set m
+    _≈ᴹ_ : Rel Carrierᴹ ℓm
+  open Str _≈ᴹ_
+  field
+    _+ᴹ_ : Op₂ Carrierᴹ
+    _*ᵣ_ : Opᵣ Carrier Carrierᴹ
+    0ᴹ : Carrierᴹ
+    -ᴹ_ : Op₁ Carrierᴹ
+    isRightModule : IsRightModule ring _+ᴹ_ _*ᵣ_ 0ᴹ -ᴹ_
+
+  open IsRightModule isRightModule public
+
+  rightSemimodule : RightSemimodule semiring m ℓm
+  rightSemimodule = record { isRightSemimodule = isRightSemimodule }
+
+------------------------------------------------------------------------
+-- Modules over commutative structures
+------------------------------------------------------------------------
+
+record Semimodule {r ℓr} (commutativeSemiring : CommutativeSemiring r ℓr) m ℓm
+                  : Set (r ⊔ ℓr ⊔ suc (m ⊔ ℓm)) where
+  open CommutativeSemiring commutativeSemiring
+
+  infixr 7 _*ₗ_
+  infixl 7 _*ᵣ_
+  infixl 6 _+ᴹ_
+  infix 4 _≈ᴹ_
+
+  field
+    Carrierᴹ : Set m
+    _≈ᴹ_ : Rel Carrierᴹ ℓm
+  open Str _≈ᴹ_
+  field
+    _+ᴹ_ : Op₂ Carrierᴹ
+    _*ₗ_ : Opₗ Carrier Carrierᴹ
+    0ᴹ : Carrierᴹ
+    isSemimodule : IsSemimodule commutativeSemiring _+ᴹ_ _*ₗ_ 0ᴹ
+
+  open IsSemimodule isSemimodule public
+
+  private
+    module L = LFP _≈_ _≈ᴹ_
+    module R = RFP _≈_ _≈ᴹ_
+
+  leftSemimodule : LeftSemimodule semiring m ℓm
+  leftSemimodule = record { isLeftSemimodule = isLeftSemimodule }
+
+  open LeftSemimodule leftSemimodule public
+    using ( +ᴹ-commutativeMonoid; +ᴹ-monoid; +ᴹ-semigroup; +ᴹ-magma
+          ; +ᴹ-rawMonoid; +ᴹ-rawMagma)
+
+  _*ᵣ_ : Opᵣ Carrier Carrierᴹ
+  _*ᵣ_ = flip _*ₗ_
+
+  *ₗ-comm : L.Commutative _*ₗ_
+  *ₗ-comm x y m =
+    M-trans (M-sym (*ₗ-assoc x y m))
+            (M-trans (*ₗ-cong (*-comm _ _) M-refl)
+                     (*ₗ-assoc y x m))
+
+  *ᵣ-comm : R.Commutative _*ᵣ_
+  *ᵣ-comm m x y = *ₗ-comm y x m
+
+  rightSemimodule : RightSemimodule semiring m ℓm
+  rightSemimodule = record
+    { Carrierᴹ = Carrierᴹ
+    ; _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
+    ; isRightSemimodule = record
+      { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+      ; *ᵣ-cong = flip *ₗ-cong
+      ; *ᵣ-zeroʳ = *ₗ-zeroˡ
+      ; *ᵣ-distribˡ = λ m x y → *ₗ-distribʳ m x y
+      ; *ᵣ-identityʳ = λ m → *ₗ-identityˡ m
+      ; *ᵣ-assoc = λ m x y → M-trans (*ₗ-comm y x m) (M-sym (*ₗ-assoc x y m))
+      ; *ᵣ-zeroˡ = *ₗ-zeroʳ
+      ; *ᵣ-distribʳ = λ x m n → *ₗ-distribˡ x m n
+      }
+    }
+
```

Comment:
If you start with `isRightSemimodule : IsRightSemimodule (...)` you should then
be able to define `rightSemimodule = record { isRightSemimodule = isRightSemimodule }`.